test/k8sT/CT.go

```diff
@@ -0,0 +1,84 @@
+// Copyright 2020 Authors of Cilium
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package k8sTest
+
+import (
+	"context"
+	"fmt"
+	"strconv"
+	"strings"
+
+	. "github.com/onsi/gomega"
+
+	. "github.com/cilium/cilium/test/ginkgo-ext"
+	"github.com/cilium/cilium/test/helpers"
+)
+
+var _ = Describe("K8sConnectionTracking", func() {
```

Comment:
I wonder, can we achieve the same testing through `bpf/tests/conntrack_test.h` `test___ct_lookup()` instead? If we're just trying to validate timeout behaviour, that's one of the few things we actually have unit test infrastructure that we can use to validate.

It should take fractions of a second to validate the change in functionality, not seconds or minutes worth of configuring the cluster, creating apps, running commands to generate traffic, just to check a timeout.

If the goal of this is to put some infrastructure in place to be able to build out more comprehensive testing, or if there's unusual application behaviour that requires integration testing to properly simulate, then I can see arguments to have additional tests like this. Otherwise I wonder if we can gain the same validation in a more lightweight manner.

Reply:
I agree that it is too much overhead for this simple use-case. I also think that the unit tests will not have the same coverage as the current test, since AFAIU we do not process actual packets in the latter.

I think there is a (mostly) happy middle ground where we:  create a tap device, load parts (e.g., the CT code with some minimal initial processing) of the datapath, send constructed packets to the interface, and check the state from the BPF maps. I think that this would generally very useful to have, and I cannot think of a reason why it wouldn't work. I want to try and hack something up and see if I can make it work.

Reply:
There's also I believe the option to load a program, then simulate running a packet through using the `BPF_PROG_TEST_RUN` command to the `bpf()` syscall.

Reply:
Pushed a new version that uses such a test.